### PR TITLE
pacific: kv/rocksdb: support migration from nautilus configured to use column families

### PR DESCRIPTION
**Pull request by IlsooByun**
Fixes: https://tracker.ceph.com/issues/57292

Signed-off-by: Ilsoo Byun <ilsoobyun@linecorp.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Review by ifed01 — APPROVED** [review]

**Comment by ifed01** [comment]
@aclamk - mind taking a look as well?

File: src/kv/RocksDBStore.cc [hunk 1]
```diff
@@ -1053,6 +1049,17 @@ int RocksDBStore::verify_sharding(const rocksdb::Options& opt,
   }
   dout(5) << __func__ << " column families from rocksdb: " << rocksdb_cfs << dendl;
 
+  if (stored_sharding_text.empty() && rocksdb_cfs.size() > 1) {
```
[the remark this answers is not in the document]
Comment (IlsooByun) [reply src/kv/RocksDBStore.cc:1052]: @ifed01 OK, I'll do it. 
To briefly explain here, sharding/def files are not created when column family is used in nautilus. However, in the current code, if there is no sharding/def file, it is assumed that only the default column family is used. 
This code allows the column family to be loaded correctly when there is no sharding/def file but actually column families are used in nautilus.

**Comment by aclamk** [comment]
I do not think I ever considered an upgrade scenario from nautilus with sharding on.

There is a code to check, but not to recreate sharding/def.
It is by choice. When a prefix is split into multiple shards (columns) it is necessary to recover the hash settings.
For example, for objects it is `O(3,0-13)` meaning that only first 13 bytes of key are used to calc hash.

In case of transit from nautilus sharding style there at max only one column per prefix and the hash does not matter at all. It seems OK to recreate it.
But there is one snag: RocksDB `repair()` function deletes all files that it does not recognize. In sunny scenario, `shaerding/def` is recreated, but if repair() breaks by sigsegv we might not have `sharding/def`. 

So, the recreation procedure must be restricted to cases when there is only 1 column per prefix.

How about an alternative solution, that will directly inject sharding/def ? :
```
echo "M= P= O=" >a_def_file
ceph-bluestore-tool --path <**osd**> bluefs-import --input-file a_def_file --dest-file sharding/def
```

As a side note, for myself, it is a disgrace that in case of broken sharding/def, a `ceph-bluestore-tool reshard` does not work (!).

**Comment by IlsooByun** [comment]
@aclamk I understood. The way you suggested may be safer.

**Comment by ljflores** [comment]
jenkins retest this please

**Comment by ljflores** [comment]
Closing, as this PR would need a rework to be included in the last point release for Pacific. Thank you for your contribution!

**Pull request closed without merging**